cmd/mdatagen/internal/samplereceiver/internal/metadata/telemetry_test.go

```diff
@@ -0,0 +1,96 @@
+	batchSizeTriggerSendKV := []attribute.KeyValue{stringAttr, stateAttrOut, enumAttr, booleanAttr}
+
+	// Ensure async metric is set upon setup.
+	ts, err := tt.GetMetric("otelcol_process_runtime_total_alloc_bytes", io_prometheus_client.MetricType_COUNTER, []attribute.KeyValue{})
+	require.NoError(t, err)
+	require.Equal(t, float64(1), ts.GetCounter().GetValue())
+
+	tb.BatchSizeTriggerSend.Add(context.Background(), 1, metric.WithAttributeSet(batchSizeTriggerSendAttrs))
+	ts, err = tt.GetMetric("otelcol_batch_size_trigger_send", io_prometheus_client.MetricType_COUNTER, batchSizeTriggerSendKV)
+	require.NoError(t, err)
+	require.Equal(t, float64(1), ts.GetCounter().GetValue())
+
+	// Ensure different attributes means resetting expected value here.
+	tb.BatchSizeTriggerSend.Add(context.Background(), 1, metric.WithAttributeSet(attribute.NewSet()))
+	ts, err = tt.GetMetric("otelcol_batch_size_trigger_send", io_prometheus_client.MetricType_COUNTER, []attribute.KeyValue{})
+	require.NoError(t, err)
+	require.Equal(t, float64(1), ts.GetCounter().GetValue())
+
+	// For histograms, use metadata buckets to set information here. Want to include values in different buckets,
+	// and then check to make sure they're in the right buckets.
+	tb.RequestDuration.Record(context.Background(), 1, metric.WithAttributeSet(attribute.NewSet(stateAttr)))
+	tb.RequestDuration.Record(context.Background(), 15, metric.WithAttributeSet(attribute.NewSet(stateAttr)))
+	ts, err = tt.GetMetric("otelcol_request_duration", io_prometheus_client.MetricType_HISTOGRAM, []attribute.KeyValue{stateAttrOut})
+	require.NoError(t, err)
+	require.Equal(t, uint64(1), *ts.GetHistogram().Bucket[0].CumulativeCount)
+	require.Equal(t, uint64(1), *ts.GetHistogram().Bucket[1].CumulativeCount)
+	require.Equal(t, uint64(2), *ts.GetHistogram().Bucket[2].CumulativeCount)
+	require.Equal(t, uint64(2), *ts.GetHistogram().Bucket[3].CumulativeCount)
+
+	// For optional metrics, we should check to make sure they are not recorded before being initialized, even if
+	// another metric has been recorded, which would trigger the callback.
+	_, err = tt.GetMetric("otelcol_queue_length", io_prometheus_client.MetricType_GAUGE, []attribute.KeyValue{})
+	require.Error(t, err)
+
+	// Init optional metric and trigger record from another metric.
+	// We can check the other metric to ensure it's correctly updated as well.
+	err = tb.InitQueueLength(func() int64 { return 1 })
+	require.NoError(t, err)
+	tb.BatchSizeTriggerSend.Add(context.Background(), 1, metric.WithAttributeSet(batchSizeTriggerSendAttrs))
+	ts, err = tt.GetMetric("otelcol_queue_length", io_prometheus_client.MetricType_GAUGE, []attribute.KeyValue{})
+	require.NoError(t, err)
+	require.Equal(t, float64(1), ts.GetGauge().GetValue())
+	ts, err = tt.GetMetric("otelcol_batch_size_trigger_send", io_prometheus_client.MetricType_COUNTER, batchSizeTriggerSendKV)
+	require.NoError(t, err)
+	require.Equal(t, float64(2), ts.GetCounter().GetValue())
```

Comment:
Just checking there were two metrics added as part of this PR but I can't seem to find a test for them

Reply:
Good call out. I'll use them in the planned follow up PR for generated testing, but they're not needed yet here. I'll remove for now. 👍